### PR TITLE
Add parameters for high pT particle protection in PUPPI

### PR DESCRIPTION
**Pull request by ahinzmann**
#### PR description:

Add new parameter to PUPPI to deal with high pT particles. Recent studies have shown that the current handling of high momentum charged particles in PUPPI is suboptimal compared to CHS and a different configuration of PUPPI for Run3 is proposed.
https://indico.cern.ch/event/862332/contributions/3632864/
This PR implements the necessary switches without changing the current behavior of the algorithm, i.e. no changes to RECO+MiniAOD are expected from this PR.

At the same time the behavior of the parameter puppiNoLeptons, that is not used anywhere in CMSSW, is adapted such that implementation actually matches the name.

#### PR validation:

runtests

#### Backports

We would like to have these options backported to 10_6_X, to introduce a new PUPPI configuration for UL MiniAOD-v2 and for JMEcustomNanoAODs as soon as the PR is merged.

### CODE REVIEW

**Comment by perrotta** [comment]
please test

File: CommonTools/PileupAlgos/plugins/PuppiProducer.cc [hunk 1]
```diff
@@ -157,13 +160,15 @@ void PuppiProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
           }
           if (tmpFromPV == 1 || tmpFromPV == 2) {
```
Comment (perrotta) [inline CommonTools/PileupAlgos/plugins/PuppiProducer.cc]: ```else if``` 
(same at line 158)
(same at line 185)
(quite likely also same at lines 189 and 192)

File: CommonTools/PileupAlgos/plugins/PuppiProducer.cc [hunk 2]
```diff
@@ -157,13 +160,15 @@ void PuppiProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
...
             pReco.id = 0;
-            if (!fPuppiForLeptons && fUseDZ && (std::abs(pDZ) < fDZCut))
+            if ((fPtMaxCharged > 0) and (pReco.pt > fPtMaxCharged))
               pReco.id = 1;
-            if (!fPuppiForLeptons && fUseDZ && (std::abs(pDZ) > fDZCut))
+            else if (fUseDZ && (std::abs(pDZ) < fDZCut))
+              pReco.id = 1;
+            else if (fUseDZ && (std::abs(pDZ) > fDZCut))
```
[the remark this answers is not in the document]
Comment (ahinzmann) [reply CommonTools/PileupAlgos/plugins/PuppiProducer.cc]: done

File: CommonTools/PileupAlgos/plugins/PuppiProducer.cc [hunk 3]
```diff
@@ -187,13 +192,15 @@ void PuppiProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
           if (lPack->fromPV() == (pat::PackedCandidate::PVTight) ||
               lPack->fromPV() == (pat::PackedCandidate::PVLoose)) {
             pReco.id = 0;
-            if (!fPuppiForLeptons && fUseDZ && (std::abs(pDZ) < fDZCut))
+            if ((fPtMaxCharged > 0) and (pReco.pt > fPtMaxCharged))
+              pReco.id = 1;
+            else if (fUseDZ && (std::abs(pDZ) < fDZCut))
```
[the remark this answers is not in the document]
Comment (ahinzmann) [reply CommonTools/PileupAlgos/plugins/PuppiProducer.cc]: done

File: CommonTools/PileupAlgos/plugins/PuppiProducer.cc (hunk 1 above)
Comment (ahinzmann) [reply CommonTools/PileupAlgos/plugins/PuppiProducer.cc]: done

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
+1
- A new parameterization is made available for puppi weights, as planned: justification for it is in the attached presentation, while a full large stat validation will follow.
- puppiNoLepton updated to better corresponds to its name
- Modified configs are not activated in the current workflows: no change in reco or miniAOD outputs onserved
- Jenkins tests pass

**Comment by santocch** [comment]
+1

**Comment by fabiocos** [comment]
+1

**Pull request merged**